src/indexable.cr

```diff
@@ -281,6 +281,66 @@ module Indexable(T)
     self
   end
 
+  # Optimized version of `Enumerable#join` that performs better when
+  # all of the elements in this indexable are strings: the total string
+  # bytesize to return can be computed before creating the final string,
+  # which performs better because there's no need to do reallocations.
+  def join(separator = "")
+    return "" if empty?
+
+    {% if T == String %}
+      _join_strings(separator)
+    {% elsif String < T %}
+      if all?(&.is_a?(String))
+        _join_strings(separator)
+      else
+        super(separator)
+      end
+    {% else %}
+      super(separator)
+    {% end %}
+  end
+
+  private def _join_strings(separator)
+    separator = separator.to_s
+
+    # The total bytesize of the string to return is:
+    length =
+      ((self.size - 1) * separator.bytesize) + # the bytesize of all separators
+        self.sum(&.to_s.bytesize)              # the bytesize of all the elements
+
+    String.new(length) do |buffer|
+      # Also compute the UTF-8 size if we can
+      size = 0
+      size_known = true
+
+      each_with_index do |elem, i|
+        # elem is guaranteed to be a String, but the compiler doesn't know this
+        # if we enter via the all?(&.is_a?(String)) branch.
+        elem = elem.to_s
+
+        # Copy separator to buffer
+        if i != 0
+          buffer.copy_from(separator.to_unsafe, separator.bytesize)
+          buffer += separator.bytesize
+        end
+
+        # Copy element to buffer
+        buffer.copy_from(elem.to_unsafe, elem.bytesize)
+        buffer += elem.bytesize
+
+        # Check whether we'll know the final UTF-8 size
+        if elem.size_known?
+          size += elem.size
```

Comment:
This is wrong because the bytesize of the seperator isn't added. The specs pass though, is there any way to test this in the specs?

Reply:
Oh, good catch! Yes, I'll check it now.

Reply:
Fixed!